File: holo-e2e-performance-tool/src/main/java/com/alibaba/hologres/performace/client/PutTest.java
package com.alibaba.hologres.performace.client;


import com.alibaba.hologres.client.HoloClient;
import com.alibaba.hologres.client.HoloConfig;
import com.alibaba.hologres.client.impl.util.ConnectionUtil;
import com.alibaba.hologres.client.model.*;
import com.alibaba.hologres.client.utils.ConfLoader;
import com.alibaba.hologres.client.utils.Metrics;
import com.alibaba.hologres.com.codahale.metrics.Histogram;
import com.alibaba.hologres.com.codahale.metrics.Meter;
import com.alibaba.hologres.org.postgresql.PGProperty;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.math.BigDecimal;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Types;
import java.util.*;
import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicLong;

public abstract class PutTest {
  public static final Logger LOG = LoggerFactory.getLogger(PutTest.class);

  protected String confName;
  protected PutTestConf conf = new PutTestConf();
  protected long targetTime;
  protected AtomicLong totalCount = new AtomicLong(0);
  protected static long memoryUsage = 0;
  protected CyclicBarrier barrier = null;
  protected HoloVersion version;



  public void run(String confName) throws Exception {
    LOG.info("confName:{}", confName);
    this.confName = confName;
    ConfLoader.load(confName, "put.", conf);
    HoloConfig config = new HoloConfig();
    ConfLoader.load(confName, "holoClient.", config);
    Properties props = new Properties();
    PGProperty.USER.set(props, config.getUsername());
    PGProperty.PASSWORD.set(props, config.getPassword());
    PGProperty.APPLICATION_NAME.set(props, config.getAppName());
    LOG.info("props : {}", props);
    String jdbcUrl = config.getJdbcUrl();
    this.init();
    Reporter reporter = new Reporter(confName);

    try (Connection conn = DriverManager.getConnection(jdbcUrl, props)) {
      if (conf.createTableBeforeRun) {
        SqlUtil.createTable(conn, conf.partition, conf.tableName, conf.columnCount,
                conf.shardCount,
                conf.orientation, conf.enableBinlog, conf.binlogTTL, conf.enableBitmap, conf.additionTsColumn,
                conf.hasPk, conf.prefixPk, conf.dataColumnType);
      }
      if (conf.enableBinlogConsumption) {
        SqlUtil.createBinLogExtension(conn);
        if (conf.recreatePublicationAndSlot) {
          SqlUtil.dropPublication(conn, conf.publicationName);
          try {
            SqlUtil.deleteReplicationProgress(conn, conf.slotName);
          } catch (SQLException e) {
            e.printStackTrace();
          }
          try {
            SqlUtil.dropSlot(conn, conf.slotName);
          } catch (SQLException e) {
            e.printStackTrace();
          }
          SqlUtil.createPublication(conn, conf.publicationName, conf.tableName);
          SqlUtil.createSlot(conn, conf.publicationName, conf.slotName);
        }
      }
      if (conf.vacuumTableBeforeRun) {
        SqlUtil.vaccumTable(conn, conf.tableName);
      }
      version = ConnectionUtil.getHoloVersion(conn);
      reporter.start(version);
    }

    barrier = new CyclicBarrier(conf.threadSize, ()->{
      memoryUsage = Util.getMemoryStat();
      Util.dumpHeap(confName);
    });
    targetTime = System.currentTimeMillis() + conf.testTime;
    Thread[] threads = new Thread[conf.threadSize];
    Metrics.startSlf4jReporter(60L, TimeUnit.SECONDS);
    for (int i = 0; i < threads.length; ++i) {
      threads[i] = new Thread(buildJob(i));
      threads[i].start();
    }

    for (int i = 0; i < threads.length; ++i) {
      threads[i].join();
    }
    LOG.info("finished, {} rows has written", totalCount.get());

    Metrics.reporter().report();
    {
      Meter meter = Metrics.registry().meter(Metrics.METRICS_WRITE_RPS);
      Histogram hist = Metrics.registry().histogram(Metrics.METRICS_WRITE_LATENCY);
      reporter.report(meter.getCount(), meter.getOneMinuteRate(), meter.getFiveMinuteRate(),
          meter.getFifteenMinuteRate(), hist.getSnapshot().getMean(),
          hist.getSnapshot().get99thPercentile(), hist.getSnapshot().get999thPercentile(), memoryUsage);
    }

    if (conf.deleteTableAfterDone) {
      SqlUtil.dropTableByHoloClient(config, conf.tableName);
    }
  }

  abstract Runnable buildJob(int id);

  abstract void init() throws Exception;

  protected void fillRecord(Record record, long pk, TableSchema schema, Random random,
                            List<String> writeColumns) {
    fillRecord(record, pk, schema, random, writeColumns, false);
  }
  protected void fillRecord(Record record, long pk, TableSchema schema, Random random,
      List<String> writeColumns, boolean enableRandomPartialCol) {
    Record put = record;
    for (String columnName : writeColumns) {
      int columnIndex = schema.getColumnIndex(columnName);
      Column column = schema.getColumn(columnIndex);
      long value = pk;
      if (!schema.isPrimaryKey(columnName) && enableRandomPartialCol) {
        int randNum = random.nextInt(3);
        if (randNum == 0) {
          continue;
        }
      }
      Set<String> distributionKeys = new HashSet<>(Arrays.asList(schema.getDistributionKeys()));
      if (conf.prefixPk && schema.isPrimaryKey(columnName) && distributionKeys.contains(columnName)) {
        value = (pk - 1) / conf.recordCountPerPrefix;
      }

      if (schema.getPartitionIndex() == columnIndex) {
        if (conf.partition) {
          value = random.nextInt(conf.partitionCount + conf.partitionRatio);
          if (value > conf.partitionCount) {
            value = 0;
          }
        } else {
          value = 0;
        }
      }
      switch (column.getType()) {
        case Types.DOUBLE:
          put.setObject(columnIndex, value);
          break;
        case Types.NUMERIC:
          put.setObject(columnIndex, new BigDecimal(value));
        case Types.INTEGER:
        case Types.BIGINT:
        case Types.SMALLINT:
          put.setObject(columnIndex, value);
          break;
        case Types.VARCHAR:
          if (columnIndex == schema.getPartitionIndex()) {
            put.setObject(columnIndex, String.valueOf(value));
          } else {
            put.setObject(columnIndex, Util.alignWithColumnSize(value, conf.columnSize));
          }
          break;
        case Types.BINARY:
          byte[] bytes = new byte[conf.columnSize];
          random.nextBytes(bytes);
          put.setObject(columnIndex, bytes);
          break;
        case Types.TIMESTAMP:
          if (conf.fillTimestampWithNow) {
            put.setObject(columnIndex, new Date());
          } else {
            put.setObject(columnIndex, new Date(value));
          }
          break;
        case Types.DATE:
          if (conf.fillTimestampWithNow) {
            put.setObject(columnIndex, new java.sql.Date(System.currentTimeMillis()));
          } else {
            put.setObject(columnIndex, new java.sql.Date(value));
          }
          break;
        default:
          throw new RuntimeException(
              "unknow type " + column.getType() + "," + column.getTypeName());
      }
    }
  }

}

class PutTestConf {
  public int threadSize = 10;
  public long testTime = 600000;
  public long rowNumber = 1000000;

  public boolean testByTime = true;

  public String tableName = "holo_perf";
  public String publicationName = "holo_perf_publication";
  public String slotName = "holo_perf_slot";
  public boolean partition = false;
  public int partitionRatio = 10;
  public int partitionCount = 30;
  public int columnCount = 100;
  public int columnSize = 10;
  public String dataColumnType = "text";
  public String orientation = "column";
  public int shardCount = -1;
  public boolean enableBinlog = false;
  public boolean enableBinlogConsumption = false;
  public boolean recreatePublicationAndSlot = true;
  public int binlogTTL = 19600; //6 hours
  public int writeColumnCount = -1;

  public boolean additionTsColumn = true;
  public boolean hasPk = true;
  public boolean fillTimestampWithNow = true;
  public boolean enableBitmap = true;

  public boolean createTableBeforeRun = true;
  public boolean vacuumTableBeforeRun = false;
  public boolean deleteTableAfterDone = true;

  public boolean prefixPk = false;
  public int recordCountPerPrefix = 100;

  public boolean dumpMemoryStat = false;

  public String beforeSql = null;

  public String afterSql = null;
}

File: holo-client-py/perf/put_test.py
"""Abstract base class for write performance tests.

Mirrors Java PutTest: handles table creation, thread management,
fill_record logic, and lifecycle.
"""

from __future__ import annotations

import logging
import random
import threading
import time
from abc import ABC, abstractmethod
from datetime import datetime, timezone
from typing import Callable, List

import psycopg

from hologres.column import (
    BIGINT,
    BINARY,
    BOOLEAN,
    DATE,
    DOUBLE,
    INTEGER,
    NUMERIC,
    REAL,
    SMALLINT,
    TIMESTAMP,
    TIMESTAMP_WITH_TIMEZONE,
    VARCHAR,
)
from hologres.config import HoloConfig
from hologres.put import Put
from hologres.table_schema import TableSchema

from .conf_loader import build_holo_config, load_conf
from .config import PutTestConf
from .metrics import Histogram, Meter
from .reporter import Reporter
from .sql_util import create_table, drop_table, vacuum_table
from .util import AtomicLong, align_with_column_size, get_write_columns

logger = logging.getLogger(__name__)


class PutTest(ABC):
    """Abstract base for write tests (INSERT / FIXED_COPY)."""

    def __init__(self):
        self.conf = PutTestConf()
        self.holo_config: HoloConfig = None
        self.conf_name = ""
        self.target_time = 0.0
        self.tic = AtomicLong(0)
        self.meter = Meter()
        self.bps_meter = Meter()
        self.histogram = Histogram()

    def run(self, conf_name: str) -> None:
        self.conf_name = conf_name
        load_conf(conf_name, "put.", self.conf)
        self.holo_config = build_holo_config(conf_name)
        self.init()

        # Setup
        with psycopg.connect(self.holo_config.conninfo, autocommit=True) as conn:
            if self.conf.create_table_before_run:
                create_table(conn, self.conf)
            if self.conf.vacuum_table_before_run:
                vacuum_table(conn, self.conf.table_name)

        reporter = Reporter(conf_name)
        reporter.start(self.holo_config.conninfo)
        self.target_time = time.time() + self.conf.test_time / 1000.0

        # Spawn threads
        threads: List[threading.Thread] = []
        for i in range(self.conf.thread_size):
            job = self.build_job(i)
            t = threading.Thread(target=job, name=f"Thread-{i}", daemon=True)
            t.start()
            threads.append(t)

        for t in threads:
            t.join()

        # Subclass cleanup (e.g. flush/close shared client)
        self.cleanup()

        logger.info("finished, %d rows written", self.meter.count)

        # Report
        reporter.report(
            self.meter, self.histogram, dump_memory_stat=self.conf.dump_memory_stat
        )

        # Cleanup
        if self.conf.delete_table_after_done:
            with psycopg.connect(self.holo_config.conninfo, autocommit=True) as conn:
                drop_table(conn, self.conf.table_name)

    def init(self) -> None:
        """Override for subclass-specific initialization."""
        pass

    def cleanup(self) -> None:
        """Override for subclass-specific cleanup after threads finish."""
        pass

    @abstractmethod
    def build_job(self, thread_id: int) -> Callable:
        """Return a callable for the thread to run."""
        ...

    def should_stop(self, row_count: int, pk: int) -> bool:
        """Check termination condition. Called every 1000 rows."""
        if self.conf.test_by_time:
            return time.time() > self.target_time
        else:
            return pk > self.conf.row_number

    def fill_record(
        self,
        put: Put,
        pk: int,
        schema: TableSchema,
        write_columns: List[str],
        rng: random.Random,
    ) -> None:
        """Populate a Put with test data matching Java fillRecord.

        Args:
            put: The Put to fill.
            pk: The primary key value.
            schema: Table schema.
            write_columns: Column names to write.
            rng: Per-thread Random instance.
        """
        for col_name in write_columns:
            col_idx = schema.get_column_index(col_name)
            if col_idx is None:
                continue
            col = schema.get_column(col_idx)

            # Determine the value for this column
            value = pk

            # For prefix PK mode
            if self.conf.prefix_pk and col_name == "id":
                value = (pk - 1) // self.conf.record_count_per_prefix
            elif self.conf.prefix_pk and col_name == "id1":
                value = pk

            # Partition column
            if col_name == "ds":
                if value >= self.conf.partition_count:
                    # Skew: values >= partitionCount map to 0
                    if self.conf.partition_ratio > 0:
                        value = rng.randint(0, self.conf.partition_count - 1)
                    else:
                        value = 0
                put.set_object(col_name, value)
                continue

            # Set by type
            col_type = col.type
            if col_type in (INTEGER, BIGINT, SMALLINT):
                put.set_object(col_name, value)
            elif col_type in (DOUBLE, REAL):
                put.set_object(col_name, float(value))
            elif col_type == NUMERIC:
                from decimal import Decimal

                put.set_object(col_name, Decimal(value))
            elif col_type == VARCHAR:
                put.set_object(
                    col_name, align_with_column_size(value, self.conf.column_size)
                )
            elif col_type == BINARY:
                put.set_object(col_name, rng.randbytes(self.conf.column_size))
            elif col_type in (TIMESTAMP, TIMESTAMP_WITH_TIMEZONE):
                if self.conf.fill_timestamp_with_now:
                    put.set_object(col_name, datetime.now(timezone.utc))
                else:
                    put.set_object(
                        col_name, datetime.fromtimestamp(value / 1000, tz=timezone.utc)
                    )
            elif col_type == DATE:
                put.set_object(
                    col_name,
                    datetime.fromtimestamp(value / 1000, tz=timezone.utc).date(),
                )
            elif col_type == BOOLEAN:
                put.set_object(col_name, value % 2 == 0)
            else:
                put.set_object(
                    col_name, align_with_column_size(value, self.conf.column_size)
                )
